pkg/storage/etcd/etcd_watcher_test.go

```diff
@@ -320,405 +283,163 @@ func makeSubsets(ip string, port int) []api.EndpointSubset {
-		}
-		if actualPod.ResourceVersion != "1" {
-			t.Errorf("Expected pod with resource version %d, Got %#v", 1, actualPod)
-		}
-		pod.ResourceVersion = "1"
-		if e, a := pod, event.Object; !api.Semantic.DeepDerivative(e, a) {
-			t.Errorf("Expected %v, got %v", e, a)
-		}
+
+	if e, a := pod, event.Object; !api.Semantic.DeepDerivative(e, a) {
+		t.Errorf("%s: expected %v, got %v", e, a)
 	}
 
-	fakeClient.WaitForWatchCompletion()
 	watching.Stop()
 }
 
-func TestWatchListIgnoresRootKey(t *testing.T) {
+func TestWatchListFromZeroIndex(t *testing.T) {
 	codec := testapi.Default.Codec()
-	pod := &api.Pod{ObjectMeta: api.ObjectMeta{Name: "foo"}}
-	key := "/some/key"
-	prefixedKey := etcdtest.AddPrefix(key)
+	key := etcdtest.AddPrefix("/some/key")
+	server := NewEtcdTestClientServer(t)
+	defer server.Terminate(t)
+	h := newEtcdHelper(server.client, codec, key)
 
-	fakeClient := tools.NewFakeEtcdClient(t)
-	h := newEtcdHelper(fakeClient, codec, etcdtest.PathPrefix())
-
-	watching, err := h.WatchList(context.TODO(), key, 1, storage.Everything)
+	watching, err := h.WatchList(context.TODO(), key, 0, storage.Everything)
 	if err != nil {
 		t.Fatalf("Unexpected error: %v", err)
 	}
-	fakeClient.WaitForWatchCompletion()
 
-	// This is the root directory of the watch, which happens to have a value encoded
-	fakeClient.WatchResponse <- &etcd.Response{
-		Action: "delete",
-		PrevNode: &etcd.Node{
-			Key:           prefixedKey,
-			Value:         runtime.EncodeOrDie(codec, pod),
-			CreatedIndex:  1,
-			ModifiedIndex: 1,
-		},
+	// creates key/foo which should trigger the WatchList for "key"
+	pod := &api.Pod{ObjectMeta: api.ObjectMeta{Name: "foo"}}
+	err = h.Create(context.TODO(), pod.Name, pod, pod, 0)
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
 	}
-	// Delete of the parent directory of a key is an event that a list watch would receive,
-	// but will have no value so the decode will fail.
-	fakeClient.WatchResponse <- &etcd.Response{
-		Action: "delete",
-		PrevNode: &etcd.Node{
-			Key:           prefixedKey,
-			Value:         "",
-			CreatedIndex:  1,
-			ModifiedIndex: 1,
-		},
+
+	event, _ := <-watching.ResultChan()
+	if event.Type != watch.Added {
+		t.Errorf("Unexpected event %#v", event)
 	}
-	close(fakeClient.WatchStop)
 
-	// the existing node is detected and the index set
-	_, open := <-watching.ResultChan()
-	if open {
-		t.Fatalf("unexpected channel open")
+	if e, a := pod, event.Object; !api.Semantic.DeepDerivative(e, a) {
+		t.Errorf("%s: expected %v, got %v", e, a)
 	}
 
 	watching.Stop()
 }
 
-func TestWatchFromNotFound(t *testing.T) {
-	fakeClient := tools.NewFakeEtcdClient(t)
-	key := "/some/key"
-	prefixedKey := etcdtest.AddPrefix(key)
-	fakeClient.Data[prefixedKey] = tools.EtcdResponseWithError{
-		R: &etcd.Response{
-			Node: nil,
-		},
-		E: &etcd.EtcdError{
-			Index:     2,
-			ErrorCode: 100,
-		},
-	}
-	h := newEtcdHelper(fakeClient, codec, etcdtest.PathPrefix())
+func TestWatchListIgnoresRootKey(t *testing.T) {
```

Comment:
B/c it ignores the changes to the rootkey it only notices changes to children of the rootkey.  

The select statement below is the validation that changes to the rootkey itself are ignored.